File: app/src/main/java/org/mifos/mobilebanking/api/services/SavingAccountsListService.java
package org.mifos.mobilebanking.api.services;

import org.mifos.mobilebanking.api.ApiEndPoints;
import org.mifos.mobilebanking.models.accounts.savings.SavingsAccountApplicationPayload;
import org.mifos.mobilebanking.models.accounts.savings.SavingsAccountUpdatePayload;
import org.mifos.mobilebanking.models.accounts.savings.SavingsAccountWithdrawPayload;
import org.mifos.mobilebanking.models.accounts.savings.SavingsWithAssociations;
import org.mifos.mobilebanking.models.payload.TransferPayload;
import org.mifos.mobilebanking.models.templates.account.AccountOptionsTemplate;
import org.mifos.mobilebanking.models.templates.savings.SavingsAccountTemplate;

import io.reactivex.Observable;
import okhttp3.ResponseBody;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.Query;

/**
 * @author Vishwajeet
 * @since 21/6/16.
 */
public interface SavingAccountsListService {

    @GET(ApiEndPoints.SAVINGS_ACCOUNTS + "/{accountId}")
    Observable<SavingsWithAssociations> getSavingsWithAssociations(
            @Path("accountId") long accountId,
            @Query("associations") String associationType);

    @GET(ApiEndPoints.ACCOUNT_TRANSFER + "/template")
    Observable<AccountOptionsTemplate> getAccountTransferTemplate();

    @POST(ApiEndPoints.ACCOUNT_TRANSFER)
    Observable<ResponseBody> makeTransfer(@Body TransferPayload transferPayload);

    @GET(ApiEndPoints.SAVINGS_ACCOUNTS + "/template")
    Observable<SavingsAccountTemplate> getSavingsAccountApplicationTemplate(
            @Query("clientId") long clientId);

    @POST(ApiEndPoints.SAVINGS_ACCOUNTS)
    Observable<ResponseBody> submitSavingAccountApplication(
            @Body SavingsAccountApplicationPayload payload);

    @PUT(ApiEndPoints.SAVINGS_ACCOUNTS + "/{accountsId}")
    Observable<ResponseBody> updateSavingsAccountUpdate(
            @Path("accountsId") String accountsId, @Body SavingsAccountUpdatePayload payload);

    @POST(ApiEndPoints.SAVINGS_ACCOUNTS + "/{savingsId}?command=withdrawnByApplicant")
    Observable<ResponseBody> submitWithdrawSavingsAccount(
            @Path("savingsId") String savingsId, @Body SavingsAccountWithdrawPayload payload);
}

File: app/src/main/java/org/mifos/mobile/api/services/SavingAccountsListService.kt
package org.mifos.mobile.api.services

import okhttp3.ResponseBody
import org.mifos.mobile.api.ApiEndPoints
import org.mifos.mobile.models.accounts.savings.SavingsAccountApplicationPayload
import org.mifos.mobile.models.accounts.savings.SavingsAccountUpdatePayload
import org.mifos.mobile.models.accounts.savings.SavingsAccountWithdrawPayload
import org.mifos.mobile.models.accounts.savings.SavingsWithAssociations
import org.mifos.mobile.models.payload.TransferPayload
import org.mifos.mobile.models.templates.account.AccountOptionsTemplate
import org.mifos.mobile.models.templates.savings.SavingsAccountTemplate
import retrofit2.Response
import retrofit2.http.*

/**
 * @author Vishwajeet
 * @since 21/6/16.
 */
interface SavingAccountsListService {
    @GET(ApiEndPoints.SAVINGS_ACCOUNTS + "/{accountId}")
    suspend fun getSavingsWithAssociations(
        @Path("accountId") accountId: Long?,
        @Query("associations") associationType: String?,
    ): SavingsWithAssociations

    @GET(ApiEndPoints.ACCOUNT_TRANSFER + "/template")
    suspend fun accountTransferTemplate(
        @Query("fromAccountId") accountId: Long?,
        @Query("fromAccountType") accountType: Long?,
    ): AccountOptionsTemplate

    @POST(ApiEndPoints.ACCOUNT_TRANSFER)
    suspend fun makeTransfer(@Body transferPayload: TransferPayload?): ResponseBody

    @GET(ApiEndPoints.SAVINGS_ACCOUNTS + "/template")
    suspend fun getSavingsAccountApplicationTemplate(
        @Query("clientId") clientId: Long?,
    ): SavingsAccountTemplate

    @POST(ApiEndPoints.SAVINGS_ACCOUNTS)
    suspend fun submitSavingAccountApplication(
        @Body payload: SavingsAccountApplicationPayload?,
    ): ResponseBody

    @PUT(ApiEndPoints.SAVINGS_ACCOUNTS + "/{accountsId}")
    suspend fun updateSavingsAccountUpdate(
        @Path("accountsId") accountsId: Long?,
        @Body payload: SavingsAccountUpdatePayload?,
    ): ResponseBody

    @POST(ApiEndPoints.SAVINGS_ACCOUNTS + "/{savingsId}?command=withdrawnByApplicant")
    suspend fun submitWithdrawSavingsAccount(
        @Path("savingsId") savingsId: String?,
        @Body payload: SavingsAccountWithdrawPayload?,
    ): ResponseBody
}
